3rdparty/everest/CMakeLists.txt

```diff
@@ -0,0 +1,40 @@
+list (APPEND everest_src)
+list (APPEND everest_inc)
+list (APPEND everest_def)
+
+execute_process(COMMAND ${PERL_EXECUTABLE} ../scripts/config.pl -f ../include/mbedtls/config.h get MBEDTLS_ECDH_VARIANT_EVEREST_ENABLED RESULT_VARIABLE result)
+
+if(${result} EQUAL 0)
+  set(everest_src
+    ${CMAKE_CURRENT_SOURCE_DIR}/library/everest.c
+    ${CMAKE_CURRENT_SOURCE_DIR}/library/x25519.c
+    ${CMAKE_CURRENT_SOURCE_DIR}/library/kremlib/FStar_UInt64_FStar_UInt32_FStar_UInt16_FStar_UInt8.c
+  )
+
+  if (${CMAKE_LIBRARY_ARCHITECTURE} STREQUAL "x86_64-linux-gnu")
```

Comment:
Anything that's not `x86_64-linux-gnu` will get the "legacy" set of sources/objects, which don't assume support for `uint128_t`. This should really have the same semantics as the `ifeq ($(shell getconf LONG_BIT),64)` in `Makefile.inc`. Is there any existing infrastructure that would allow us to check/configure this type of property?

Reply:
As noted in https://github.com/ARMmbed/mbedtls/pull/2073#discussion_r255572348 we never do any detection of the build host because most of our users cross-compile. Our current “solution” is to compile every file but to have preprocessor guards that can cause an object file to be empty (i.e. we'd put `#if MBEDTLS_HAVE_INT64` at the top of 64-bit-specific source files and `#if MBEDTLS_HAVE_INT32` at the top of 32-bit-specific source files), which is not great.

Reply:
So, can I assume that exactly one of `-DMBEDTLS_HAVE_INT32` or `-DMBEDTLS_HAVE_INT64` is always present at build time so that I can grep for it in the `CFLAGS`? I suspect that `-DMBEDTLS_HAVE_INT64` + Windows/VC still won't support uint128 in some versions though.

Reply:
No. Despite the name which looks like it's from autoconf's naming scheme, `MBEDTLS_HAVE_xxx` is not defined by the build system, it's defined in a header.

I still can't think of a better solution than what we're doing inside the library, which is to build all the files unconditionally but surround the file contents with `#if` clauses. For example, if you have a version that depends on the availability of `uint128_t`, write a header with something like
```
#include <limits.h>
#if ULLONG_MAX > 0xffffffffffffffff
#define MBEDTLS_HAVE_INT128
#endif
```
and then put `#if defined(MBEDTLS_HAVE_INT128)` around the whole source file that requires `uint128_t`.

Reply:
That would require modification of the automatically produced source files, possible jeopardizing the verification guarantees, so I'd rather not do that if there's a different way. This is mainly about chosing the right `Hacl_Curve25519.c`, I could add a new file that very unconventionally `#include`s the `.c` file instead of a `.h`, depending on the macro. Would that be an acceptable solution?

Reply:
It's ugly, but well, it works, so if we can't think of a better solution, let's do that.

(For us to change our build system to choose what to build based on target platform capabilities would be better to solve this particular problem, but it's difficult given that we want to be compatible with a wide variety of toolchains and a solution that doesn't work for cross-compilation would not be acceptable.)

Reply:
Done! 32c81ceb6c77a4409f2c44dea6a124bf9b0c5d9a